### DISCUSSION TITLE
Detect soft vs. hard navigation

### DISCUSSION THREAD

**Discussion by jamesvclements** [vercel/next.js#49824] [Help]
### Summary

Is there a way to detect a soft vs. hard navigation for a route? Or is there a better way to handle this usecase?

I'm looking to have a page load animation when someone first visits the site, but the animation shouldn't replay when they navigate back to the page. For example:
1. Initial visit to /home plays load animation
2. Visit to /about plays load animation on about page
3. Visit back to /home does not play load animation

**Comment by MilanLund** [comment]
In terms of soft vs. hard navigation, I came up with the following function: 
```
function getNavigationMode(): 'soft' | 'hard' {
    const headersList = headers();
    // If there is a next-url header, soft navigation has been performed
    // Otherwise, hard navigation has been performed
    const nextUrl = headersList.get('next-url');
    if (nextUrl) {
        return 'soft';
    }
    return 'hard';
}
```
It basically checks for the existence of the `next-url` header. I found that out by simply comparing headers for soft and hard navigation and looking for differences. I am not sure if that would work for all versions. However in  `App directory` version `13.4.12`, it does the job I need.

**Comment by johanaarstein** [comment]
One problem with the approach suggested by @MilanLund is that you lose static rendering of each page where you use `next/headers`. If this is an issue another way to achieve the same effect would be to place the loading animation in a global client component, and switch between showing it or not with a `useEffect`-hook. This way you can have unique animations for each page you soft navigate too, the first time they load.

However, @MilanLund's function is great for middleware if you want to, say, redirect visitors based on their preferred language, but not make it impossible to switch language manually.

An even simpler version is:

```javascript
import { headers } from 'next/headers'

isHardNavigation = () => {
  return !headers().has('next-url')
}
```

**Reply by AnweshGangula** [reply]
@johanaarstein, I'm trying to implement this. Can you help me understand how to define a global client component?

**Reply by johanaarstein** [reply]
Sure –  just create a client component, and name it whatever you want – say GlobalClient.tsx.
```typescript
'use client'
import {
  useEffect,
  useRef
} from 'react'
import { usePathname } from 'next/navigation'

export default function GlobalClient() {
const initialRender = useRef(true),
   pathname = usePathname()
useEffect(() => {
    if (!initialRender.current) {
      softNavTransition()
    } else {
       hardNavTransition()
       initialRender.current = false
    }
  }, [pathname])
}
```

Then place the component inside RootLayout before the closing `body` tag. This way you get to differentiate between hard and soft navigation while keeping static rendering. You can even have different transitions for different pathnames if you want.

**Comment by AnweshGangula** [comment]
For future readers and my own reference, 

If your intention is to transition content with soft navigation. Check out view transitions API
* https://github.com/vercel/next.js/discussions/42658
* https://developer.chrome.com/docs/web-platform/view-transitions/

**Comment by realamirhe** [comment]
Based on the documentation and current research:

Wrapping a component in Suspense will cause it to render a fallback until the entire boundary resolves, unless additional nested Suspense boundaries are used to allow partial streaming. When loading.tsx is present, the corresponding Server Component route will render the loading state on initial render, including hard refreshes.

Although modern search engines are capable of parsing and streaming the response incrementally, using loading.tsx means the primary content is not sent immediately in the initial HTML. From an SEO perspective, this is suboptimal, since the content is not available to crawlers at first paint, even though it may arrive shortly after via streaming.

The goal is to:

- Ensure content is fully present (or e.g. prefetched in react-query)during hard navigations for SEO, and
- Show loading states only during soft navigations to improve perceived performance,

then differentiating between hard and soft navigation becomes important. Detecting this distinction allows deferring loading states for client-side transitions while keeping the server-rendered HTML complete on hard loads.


```ts
// navigation-intent.server.ts
import { headers } from "next/headers";

/**
 * Best-effort detection of client-side navigation (SPA-style).
 *
 * There is no 100% reliable server-side signal.
 * combine multiple weak indicators and accept probabilistic truth.
 */
export async function isClientSideNavigation(): Promise<boolean> {
  const h = await headers();

  const referer = h.get("referer");
  const nextRouterPrefetch = h.get("next-router-prefetch");
  const nextRouterStateTree = h.get("next-router-state-tree");
  const secFetchDest = h.get("sec-fetch-dest");
  const secFetchMode = h.get("sec-fetch-mode");

  // Referer usually exists on soft navigations (note: it might also come from the google too)
  if (referer) return true;

  // Next.js internal router headers (most reliable in-app signal)
  if (nextRouterPrefetch || nextRouterStateTree) return true;

  // Document navigations tend to have `sec-fetch-dest: document`
  // Client-side data/navigation requests often do not (comes with 'empty')
  if (secFetchDest && secFetchDest !== "document") return true;

  // CORS requests are usually not full page navigations
  if (secFetchMode === "cors") return true;

  return false;
}
````


```tsx
// ConditionalSuspense.server.tsx
import { Suspense } from "react";
import { isClientSideNavigation } from "./navigation-intent.server";

export async function ConditionalSuspense(
  props: React.ComponentProps<typeof Suspense>
) {
  const isSoftNavigation = await isClientSideNavigation();

  if (isSoftNavigation) {
    return (
      <Suspense fallback={props.fallback}>
        {props.children}
      </Suspense>
    );
  }

  // Hard navigation: render content directly
  return <>{props.children}</>;
}
```


```tsx
// <path>/page.tsx
export default async function Page() {
  return (
    <ConditionalSuspense fallback={<Loading />}>
      <SamplePageContent />
    </ConditionalSuspense>
  );
}

async function SamplePageContent() {
  await new Promise((r) => setTimeout(r, 2000)); // Simulated async work
  return <div>CHILD</div>;
}
```



* the solution is a heuristic-based but has been tested with latest next `"16.0.8"` and it works.
* it may be missing for bots, privacy-focused browsers, or certain refresh scenarios
* This pattern should be applied **selectively**, not as a global default
* Overusing conditional Suspense can complicate mental models and debugging
